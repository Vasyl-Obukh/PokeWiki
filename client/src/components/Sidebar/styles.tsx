import styled from 'styled-components';

export const Container = styled.aside`
  position: sticky;
  padding: 25px 10px;
  height: 500px;
  border-radius: 25px;
  box-shadow: 0 0 5px darkslategray;
`;