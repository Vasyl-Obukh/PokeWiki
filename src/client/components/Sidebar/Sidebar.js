import React from 'react'
import styled from 'styled-components';

const SidebarContainer = styled.div`
  padding: 25px 10px;
  height: auto;
  border-radius: 25px;
  box-shadow: 0 0 5px darkslategray;
`;

const Sidebar = () => {
  return <SidebarContainer></SidebarContainer>;
};

export default Sidebar;