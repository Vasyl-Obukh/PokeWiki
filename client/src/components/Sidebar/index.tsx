import * as React from 'react';
import { withRouter } from 'react-router-dom';
import FilterForm from '../FilterForm';
import * as Styles from './styles';

export const normalizeSearchParams = (
  searchParams: URLSearchParams,
  {types, evoLevels}: {types: string[], evoLevels: string[]}
): URLSearchParams => {
  searchParams.set('page', '1');
  types.length ?
    searchParams.set('types', types.toString())
    : searchParams.delete('types');
  evoLevels.length ?
    searchParams.set('evoLevels',  evoLevels.toString())
    : searchParams.delete('evoLevels');
  return searchParams;
};

type FormValues = {
  [key: string]: boolean
};

export const convertToStringArray = (values: FormValues): string[] =>
  Object.entries(values)
    .filter(_ => _[1])
    .map(_ => _[0]);

const Sidebar = (props) => {
  const handleSubmit = (values: {[key: string]: FormValues}): void => {
    const searchParams: URLSearchParams = new URLSearchParams(props.location.search);

    const types: string[] = convertToStringArray(values.types);
    const evoLevels: string[] = convertToStringArray(values.evoLevels);

    const normalizedParams: URLSearchParams  = normalizeSearchParams(searchParams, {types, evoLevels});

    props.history.push(`/?${normalizedParams}`);
  };

  return (
    <Styles.Container>
      <FilterForm onSubmit={handleSubmit} search={props.location.search} />
    </Styles.Container>
  );
};

export default withRouter(Sidebar);