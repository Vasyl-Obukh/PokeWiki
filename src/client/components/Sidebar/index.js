import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import FilterForm from '../FilterForm';
import * as Styles from './styles';

export const normalizeSearchParams = (searchParams, {types, evoLevels}) => {
  searchParams.set('page', 1);
  types.length ?
    searchParams.set('types', types)
    : searchParams.delete('types');
  evoLevels.length ?
    searchParams.set('evoLevels',  evoLevels)
    : searchParams.delete('evoLevels');
  return searchParams;
};

export const convertToStringArray = values =>
  Object.entries(values)
    .filter(_ => _[1])
    .map(_ => _[0]);

const Sidebar = (props) => {
  const handleSubmit = values => {
    const searchParams = new URLSearchParams(props.location.search);

    const types = convertToStringArray(values.types);
    const evoLevels = convertToStringArray(values.evoLevels);

    const normalizedParams  = normalizeSearchParams(searchParams, {types, evoLevels});

    props.history.push(`/?${normalizedParams}`);
  };

  return (
    <Styles.Container>
      <FilterForm onSubmit={handleSubmit} search={props.location.search} />
    </Styles.Container>
  );
};

Sidebar.propTypes = {
  history: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired
};

export default withRouter(Sidebar);