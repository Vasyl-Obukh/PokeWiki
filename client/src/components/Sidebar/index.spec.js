import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Sidebar, { normalizeSearchParams, convertToStringArray } from './index';

describe('<Sidebar /> tests', () => {
  const mockStore = configureStore();
  const initialState = {};
  const store = mockStore(initialState);
  let props;

  beforeEach(() => {
    props = {search: {location: {search: 'page=2&&types=normal,fire'}}};
  });

  test('Should render correctly', () => {
    const component = renderer.create(
      <Provider store={store}>
        <Router>
          <Sidebar {...props}/>
        </Router>
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('normalizeSearchParams() tests', () => {

  test('Should change all keys', () => {
    const searchParams = new URLSearchParams('page=3&&types=fire&&evoLevels=2');
    const update = {
      types: ['water'],
      evoLevels: [1, 3]
    };
    expect(normalizeSearchParams(searchParams, update))
      .toEqual(new URLSearchParams('page=1&&types=water&&evoLevels=1,3'));
  });

  test('Should delete all keys except page', () => {
    const searchParams = new URLSearchParams('page=3&&types=fire&&evoLevels=2');
    const update = {
      types: [],
      evoLevels: []
    };
    expect(normalizeSearchParams(searchParams, update))
      .toEqual(new URLSearchParams('page=1'));
  });
});

describe('convertToStringArray() tests', () => {
  test('Should work correctly', () => {
    const values = {fire: true, electric: false, normal: true};
    expect(convertToStringArray(values)).toEqual(['fire', 'normal']);
  });
});