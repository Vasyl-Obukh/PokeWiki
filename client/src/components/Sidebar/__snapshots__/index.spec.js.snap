// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> tests Should render correctly 1`] = `
<aside
  className="sc-ifAKCX fYquXR"
>
  <form
    className="sc-bdVaJa fUGKAz"
    onSubmit={[Function]}
  >
    <h2>
      Filter by:
    </h2>
    <h3>
      Types
    </h3>
    <div
      className="sc-bwzfXH gTUROU"
    >
      <div>
        <input
          checked={false}
          id="normal"
          name="types.normal"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="normal"
        >
          normal
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="fighting"
          name="types.fighting"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="fighting"
        >
          fighting
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="flying"
          name="types.flying"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="flying"
        >
          flying
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="poison"
          name="types.poison"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="poison"
        >
          poison
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="ground"
          name="types.ground"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="ground"
        >
          ground
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="rock"
          name="types.rock"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="rock"
        >
          rock
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="bug"
          name="types.bug"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="bug"
        >
          bug
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="ghost"
          name="types.ghost"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="ghost"
        >
          ghost
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="steel"
          name="types.steel"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="steel"
        >
          steel
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="fire"
          name="types.fire"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="fire"
        >
          fire
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="water"
          name="types.water"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="water"
        >
          water
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="grass"
          name="types.grass"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="grass"
        >
          grass
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="electric"
          name="types.electric"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="electric"
        >
          electric
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="psychic"
          name="types.psychic"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="psychic"
        >
          psychic
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="ice"
          name="types.ice"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="ice"
        >
          ice
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="dragon"
          name="types.dragon"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="dragon"
        >
          dragon
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="fairy"
          name="types.fairy"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="fairy"
        >
          fairy
        </label>
      </div>
    </div>
    <h3>
      Evolution levels
    </h3>
    <div
      className="sc-bwzfXH gTUROU"
    >
      <div>
        <input
          checked={false}
          id="1"
          name="evoLevels.1"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="1"
        >
          1
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="2"
          name="evoLevels.2"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="2"
        >
          2
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="3"
          name="evoLevels.3"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="3"
        >
          3
        </label>
      </div>
      <div>
        <input
          checked={false}
          id="4"
          name="evoLevels.4"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          type="checkbox"
          value=""
        />
        <label
          className="sc-htpNat eWCEFP"
          htmlFor="4"
        >
          4
        </label>
      </div>
    </div>
    <button
      className="sc-bxivhb jAoLPd"
      type="submit"
    >
      filter
    </button>
  </form>
</aside>
`;
